doc-content/drools-docs/src/main/asciidoc/DecisionEngine/inference-and-truth-maintenance.adoc

```diff
@@ -0,0 +1,273 @@
+|Issue ID Card{set:cellbgcolor:#ffff99}
+
+|Issue ID Card to Adults{set:cellbgcolor:#ffffff}
+|London
+|18
+|p
+
+|===
+
+
+However the ID department does not set the policy on who an adult is.
+That's done at a central government level.
+If the central government were to change that age to 21, this would initiate a change management process.
+Someone would have to liaise with the ID department and make sure their systems are updated, in time for the law going live.
+
+This change management process and communication between departments is not ideal for an agile environment, and change becomes costly and error prone.
+Also the card department is managing more information than it needs to be aware of with its "monolithic" approach to rules management which is "leaking" information better placed elsewhere.
+By this I mean that it doesn't care what explicit "age >= 18" information determines whether someone is an adult, only that they are an adult.
+
+In contrast to this, let's pursue an approach where we split (de-couple) the authoring responsibilities, so that both the central government and the ID department maintain their own rules.
+
+It's the central government's job to determine who is an adult.
+If they change the law they just update their central repository with the new rules, which others use:
+
+
+[cols="3*^"]
+|===
+
+|{set:cellbgcolor:#bfbfbf}
+2+|[white]#RuleTable Age Policy{set:cellbgcolor:#000000}#
+
+|{set:cellbgcolor:#bfbfbf}
+|CONDITION{set:cellbgcolor:#ffcc94}
+|ACTION
+
+|{set:cellbgcolor:#bfbfbf}
+|p : Person{set:cellbgcolor:#ffcc94}
+
+|
+|{set:cellbgcolor:#bfbfbf}
+|age >= $1{set:cellbgcolor:#ffcc94}
+|insert($1)
+
+|{set:cellbgcolor:#b9ffca}
+|Adult Age Policy{set:cellbgcolor:#ccffff}
+|Add Adult Relation{set:cellbgcolor:#f7ff92}
+|Infer Adult{set:cellbgcolor:#ffffff}
+
+|18
+|new IsAdult( p )
+
+|===
+
+
+
+The IsAdult fact, as discussed previously, is inferred from the policy rules.
+It encapsulates the seemingly arbitrary piece of logic "age >= 18" and provides semantic abstractions for its meaning.
+Now if anyone uses the above rules, they no longer need to be aware of explicit information that determines whether someone is an adult or not.
+They can just use the inferred fact:
+
+
+[cols="4*^"]
+|===
+
+|{set:cellbgcolor:#c0c0c0}
+3+|[white]#RuleTable ID Card{set:cellbgcolor:#000000}#
+
+|{set:cellbgcolor:#c0c0c0}
+|CONDITION{set:cellbgcolor:#ffcc99}
+|CONDITION
+|ACTION
+
+|{set:cellbgcolor:#c0c0c0}
+|p : Person{set:cellbgcolor:#ffcc99}
+|isAdult
+|
+
+|{set:cellbgcolor:#c0c0c0}
+|location{set:cellbgcolor:#ffcc99}
+|person == $1
+|issueIdCard($1)
+
+|{set:cellbgcolor:#ccffcc}
+|Select Person{set:cellbgcolor:#ccffff}
+|Select Adults
+|Issue ID Card{set:cellbgcolor:#ffff99}
+
+|Issue ID Card to Adults{set:cellbgcolor:#ffffff}
+|London
+|p
+|p
+|
+
+|===
+
+
+While the example is very minimal and trivial it illustrates some important points.
+We started with a monolithic and leaky approach to our knowledge engineering.
```

Comment:
Should these be bullet pointed?

Reply:
This section is conditionalized for community only because it's not worth retaining and polishing, but  Jozef Marko put a lot of work into the tables, so I wanted to at least leave it for community. Consequently, what you see here (believe it or not) is a recommended ascidoc text formatting, like in coding, where each new sentence is a new line. It renders as a string/paragraph in the output until you enter two new lines (one with white space). The value is that git addresses conflicts per line, so the longer the line (i.e., paragraph) the more vulnerable to conflicts it becomes. If I change word 1 in the paragraph and you change word 37, a conflict arises because technically the same "line" in asciidoc. By breaking lines at each sentence, it avoids this.

However :) , that also makes the source quite ugly and also vulnerable to formatting issues in the output that you might not notice in the source because it's written at the sentence level rather than paragraph level. So I/we on our team don't typically do this, and I don't think we should.

Long-winded response, but I thought you should know.

Reply:
Thank you for the explanation. That makes sense.